File: ios/Runner/SuperwallDonationBridge.swift
import Flutter
import Foundation
import StoreKit
import SuperwallKit

/// Direct purchases for one-time donations.
/// Prefers SuperwallKit `products(for:)` + `purchase(_:)` when Superwall is
/// configured. Falls back to StoreKit 2 with the same product IDs so simulator
/// / StoreKit-config testing works even if the Superwall API key is missing.
/// Never presents a paywall and must never grant subscription entitlements.
enum SuperwallDonationBridge {
  static let channelName = "com.app.deenly.deenly/superwall_donations"

  static func register(messenger: FlutterBinaryMessenger) {
    let channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    channel.setMethodCallHandler { call, result in
      switch call.method {
      case "fetchProduct":
        fetchProduct(call: call, result: result)
      case "purchase":
        purchase(call: call, result: result)
      default:
        result(FlutterMethodNotImplemented)
      }
    }
  }

  private static func productId(from call: FlutterMethodCall) -> String? {
    guard let args = call.arguments as? [String: Any] else { return nil }
    let raw = (args["productId"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
    return (raw?.isEmpty == false) ? raw : nil
  }

  private static func fetchProduct(
    call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) {
    guard let productId = productId(from: call) else {
      result(
        FlutterError(
          code: "INVALID_ARGS",
          message: "productId is required",
          details: nil
        )
      )
      return
    }

    Task {
      do {
        let payload = try await fetchProductPayload(productId: productId)
        NSLog("[DeenFocus][Donation] fetched product %@", payload.description)
        DispatchQueue.main.async { result(payload) }
      } catch let error as DonationBridgeError {
        NSLog("[DeenFocus][Donation] fetch error %@", error.flutterError.message ?? "")
        DispatchQueue.main.async { result(error.flutterError) }
      } catch {
        NSLog("[DeenFocus][Donation] fetch failed %@", error.localizedDescription)
        DispatchQueue.main.async {
          result(
            FlutterError(
              code: "FETCH_FAILED",
              message: error.localizedDescription,
              details: nil
            )
          )
        }
      }
    }
  }

  private static func purchase(
    call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) {
    guard let productId = productId(from: call) else {
      result(
        FlutterError(
          code: "INVALID_ARGS",
          message: "productId is required",
          details: nil
        )
      )
      return
    }

    Task {
      do {
        let payload = try await purchaseProduct(productId: productId)
        NSLog("[DeenFocus][Donation] purchase result %@", payload.description)
        DispatchQueue.main.async { result(payload) }
      } catch let error as DonationBridgeError {
        NSLog("[DeenFocus][Donation] purchase error %@", error.flutterError.message ?? "")
        DispatchQueue.main.async { result(error.flutterError) }
      } catch {
        NSLog("[DeenFocus][Donation] purchase failed %@", error.localizedDescription)
        DispatchQueue.main.async {
          result(
            FlutterError(
              code: "PURCHASE_FAILED",
              message: error.localizedDescription,
              details: nil
            )
          )
        }
      }
    }
  }

  private static var isSuperwallReady: Bool {
    Superwall.isInitialized && Superwall.shared.configurationStatus == .configured
  }

  private static func purchaseProduct(productId: String) async throws -> [String: Any] {
    if isSuperwallReady, let product = await superwallProduct(productId: productId) {
      NSLog("[DeenFocus][Donation] purchasing via Superwall id=%@", productId)
      let purchaseResult = await Superwall.shared.purchase(product)
      switch purchaseResult {
      case .purchased:
        return ["status": "purchased", "productId": productId]
      case .cancelled:
        return ["status": "cancelled", "productId": productId]
      case .pending:
        return ["status": "pending", "productId": productId]
      case .failed(let error):
        NSLog(
          "[DeenFocus][Donation] Superwall purchase failed, trying StoreKit 2: %@",
          error.localizedDescription
        )
      }
    } else {
      NSLog(
        "[DeenFocus][Donation] Superwall not configured, using StoreKit 2 id=%@",
        productId
      )
    }

    if #available(iOS 15.0, *) {
      return try await purchaseWithStoreKit2(productId: productId)
    }
    throw DonationBridgeError.productNotFound(productId)
  }

  private static func fetchProductPayload(productId: String) async throws -> [String: Any] {
    if isSuperwallReady, let product = await superwallProduct(productId: productId) {
      return [
        "productId": product.productIdentifier,
        "localizedPrice": product.localizedPrice,
      ]
    }

    if #available(iOS 15.0, *) {
      let product = try await storeKit2Product(productId: productId)
      return [
        "productId": product.id,
        "localizedPrice": product.displayPrice,
      ]
    }

    throw DonationBridgeError.productNotFound(productId)
  }

  private static func superwallProduct(productId: String) async -> StoreProduct? {
    let products = await Superwall.shared.products(for: Set([productId]))
    return products.first(where: { $0.productIdentifier == productId })
  }

  @available(iOS 15.0, *)
  private static func storeKit2Product(productId: String) async throws -> StoreKit.Product {
    let products = try await StoreKit.Product.products(for: [productId])
    if let match = products.first(where: { $0.id == productId }) {
      return match
    }
    throw DonationBridgeError.productNotFound(productId)
  }

  @available(iOS 15.0, *)
  private static func purchaseWithStoreKit2(productId: String) async throws -> [String: Any] {
    let product = try await storeKit2Product(productId: productId)
    NSLog("[DeenFocus][Donation] StoreKit 2 sheet for %@", productId)
    let result = try await product.purchase()
    switch result {
    case .success(let verification):
      switch verification {
      case .verified(let transaction):
        await transaction.finish()
        return ["status": "purchased", "productId": productId]
      case .unverified(_, let error):
        throw error
      }
    case .userCancelled:
      return ["status": "cancelled", "productId": productId]
    case .pending:
      return ["status": "pending", "productId": productId]
    @unknown default:
      throw DonationBridgeError.purchaseFailed("Unknown StoreKit purchase result")
    }
  }
}

private enum DonationBridgeError: Error {
  case productNotFound(String)
  case purchaseFailed(String)

  var flutterError: FlutterError {
    switch self {
    case .productNotFound(let productId):
      return FlutterError(
        code: "PRODUCT_NOT_FOUND",
        message: "StoreKit product not found: \(productId)",
        details: productId
      )
    case .purchaseFailed(let message):
      return FlutterError(
        code: "PURCHASE_FAILED",
        message: message,
        details: nil
      )
    }
  }
}

File: android/app/src/main/kotlin/com/rnr/deenfocus/SuperwallDonationBridge.kt
package com.rnr.deenfocus

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.superwall.sdk.Superwall
import com.superwall.sdk.config.models.ConfigurationStatus
import com.superwall.sdk.delegate.PurchaseResult
import com.superwall.sdk.store.abstractions.product.StoreProduct
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Direct purchases for one-time donations on Android.
 *
 * Mirrors iOS: Superwall `getProducts` + `purchase` only. No app-owned
 * [com.android.billingclient.api.BillingClient] — Superwall's SDK owns the
 * Play Billing lifecycle (query, launch, consume/finish). Never presents a
 * paywall and must never grant subscription entitlements.
 */
class SuperwallDonationBridge : MethodChannel.MethodCallHandler {
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "fetchProduct" -> {
                val productId = productId(call, result) ?: return
                runAsync(result) { fetchProductPayload(productId) }
            }
            "purchase" -> {
                val productId = productId(call, result) ?: return
                runAsync(result) { purchaseProduct(productId) }
            }
            else -> result.notImplemented()
        }
    }

    private fun productId(call: MethodCall, result: MethodChannel.Result): String? {
        val raw = call.argument<String>("productId")?.trim().orEmpty()
        if (raw.isEmpty()) {
            result.error("INVALID_ARGS", "productId is required", null)
            return null
        }
        if (raw !in ALLOWED_PRODUCT_IDS) {
            result.error("PRODUCT_NOT_FOUND", "Donation product not allowed: $raw", raw)
            return null
        }
        return raw
    }

    private fun runAsync(
        result: MethodChannel.Result,
        block: suspend () -> Map<String, Any>,
    ) {
        CoroutineScope(Dispatchers.Main.immediate).launch {
            try {
                deliverSuccess(result, block())
            } catch (error: DonationBridgeException) {
                Log.e(TAG, "Donation bridge error ${error.code}: ${error.message}")
                deliverError(result, error.code, error.message, error.details)
            } catch (error: Exception) {
                Log.e(TAG, "Donation bridge failed", error)
                deliverError(
                    result,
                    "PURCHASE_FAILED",
                    error.message ?: "Donation purchase failed",
                    null,
                )
            }
        }
    }

    private suspend fun fetchProductPayload(productId: String): Map<String, Any> {
        val product = requireSuperwallProduct(productId)
        return mapOf(
            "productId" to product.productIdentifier,
            "localizedPrice" to product.localizedPrice,
        )
    }

    private suspend fun purchaseProduct(productId: String): Map<String, Any> {
        val product = requireSuperwallProduct(productId)
        Log.i(TAG, "purchasing via Superwall id=$productId")

        val outcome = Superwall.instance.purchase(product)
        val purchaseResult = outcome.getOrElse { error ->
            throw DonationBridgeException(
                "PURCHASE_FAILED",
                error.message ?: "Superwall purchase failed",
                null,
            )
        }

        return when (purchaseResult) {
            is PurchaseResult.Purchased -> statusPayload("purchased", productId)
            is PurchaseResult.Cancelled -> statusPayload("cancelled", productId)
            is PurchaseResult.Pending -> statusPayload("pending", productId)
            is PurchaseResult.Failed -> throw DonationBridgeException(
                "PURCHASE_FAILED",
                purchaseResult.errorMessage,
                null,
            )
        }
    }

    private suspend fun requireSuperwallProduct(productId: String): StoreProduct {
        if (!isSuperwallReady) {
            throw DonationBridgeException(
                "PRODUCT_NOT_FOUND",
                "Superwall is not configured",
                productId,
            )
        }

        val products = try {
            Superwall.instance.getProducts(productId).getOrElse { error ->
                throw DonationBridgeException(
                    "PRODUCT_NOT_FOUND",
                    error.message ?: "Superwall getProducts failed",
                    productId,
                )
            }
        } catch (error: DonationBridgeException) {
            throw error
        } catch (error: Exception) {
            throw DonationBridgeException(
                "PRODUCT_NOT_FOUND",
                error.message ?: "Superwall getProducts failed",
                productId,
            )
        }

        return products[productId]
            ?: products.values.firstOrNull { it.productIdentifier == productId }
            ?: throw DonationBridgeException(
                "PRODUCT_NOT_FOUND",
                "Superwall product not found: $productId",
                productId,
            )
    }

    private val isSuperwallReady: Boolean
        get() {
            if (!Superwall.initialized) {
                Log.w(TAG, "Superwall not ready: initialized=false")
                return false
            }
            return try {
                val state = Superwall.instance.configurationState
                val ready = state == ConfigurationStatus.Configured
                if (!ready) {
                    Log.w(TAG, "Superwall not ready: initialized=true state=$state")
                }
                ready
            } catch (error: Exception) {
                Log.w(TAG, "Superwall readiness check failed: ${error.message}")
                false
            }
        }

    private fun statusPayload(status: String, productId: String): Map<String, Any> {
        return mapOf("status" to status, "productId" to productId)
    }

    private fun deliverSuccess(result: MethodChannel.Result, payload: Map<String, Any>) {
        mainHandler.post { result.success(payload) }
    }

    private fun deliverError(
        result: MethodChannel.Result,
        code: String,
        message: String?,
        details: Any?,
    ) {
        mainHandler.post { result.error(code, message, details) }
    }

    private class DonationBridgeException(
        val code: String,
        override val message: String?,
        val details: Any?,
    ) : Exception(message)

    companion object {
        const val CHANNEL = "com.app.deenly.deenly/superwall_donations"

        val ALLOWED_PRODUCT_IDS: Set<String> = setOf(
            "com.deenfocus.donation.10",
            "com.deenfocus.donation.25",
            "com.deenfocus.donation.50",
            "com.deenfocus.donation.100",
            "com.deenfocus.donation.250",
        )

        private const val TAG = "DeenFocusDonation"
    }
}
